# Bank Marketing dataset

The data is related with marketing campaigns of a Portuguese banking institution that were based on phone calls. The dataset contain several attributes such as bank clients data, information about current campaign and social and economic context attributes. The classification goal is to predict whether the client will subscribe a bank term deposit.  

In [25]:
# Importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [26]:
# Setting plotting parameters
parameters = {'figure.figsize':(10, 7),
             'axes.labelsize': 11,
             'axes.titlesize':16}
plt.rcParams.update(parameters)
sns.set_style('darkgrid')

In [27]:
data = pd.read_csv('bank-additional-full.csv', sep = ';')
data

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

There are 41.188 observations and 21 variables in data including the target variable y.

### Variables description

#### Bank client data:
1. age

2. job : type of job

3. marital : marital status

4. education

5. default: has credit in default?

6. housing: has housing loan? 

7. loan: has personal loan? 

#### Related with the last contact of the current campaign:
8. contact: contact communication type

9. month: last contact month of year

10. day_of_week: last contact day of the week

11. duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.

#### Other attributes:
12. campaign: number of contacts performed during this campaign and for this client

13. pdays: number of days that passed by after the client was last contacted from a previous campaign (999 means client was not previously contacted)

14. previous: number of contacts performed before this campaign and for this client

15. poutcome: outcome of the previous marketing campaign

#### Social and economic context attributes
16. emp.var.rate: employment variation rate - quarterly indicator

17. cons.price.idx: consumer price index - monthly indicator 

18. cons.conf.idx: consumer confidence index - monthly indicator 

19. euribor3m: euribor 3 month rate - daily indicator 
   - it is calculated by eliminating the highest 15% and the lowest 15% of the interest rates submitted and calculating the
     arithmetic mean of the remaining values

20. nr.employed: number of employees - quarterly indicator 

#### Target variable:
21. y - has the client subscribed a term deposit? 

In [28]:
# Descriptive statistics for numeric data
data.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

The contacted clients were 40 years old in average. The average duration of communication is approximately 258 seconds (4 minutes, 3 seconds). The maximum value is 4918 seconds that could be an outlier (almost 82 minutes long communication). As we saw in the description of varaibles, duration variable is highly correlated with the target, so in case of prediting outcomes, we should exclude it. Variable campaign holds the number of contacts performed during this campaign. 75% of clients were contacted 3 times or less, and maximum value of contacts with a client is 56 times. Seems like the employees of the bank have tried to contact the client several times until they get the response. Variable pdays describe the number of day that passed by after the last contact with the client regarding previous campaign. 75% clients were not contacted in the previous campaign (999 = no previous contact). The similar result is for variable previous, that the majority of clients was not contacted. 

Variable emp.var.rate is indicator of employment rate calculated every quarter. The employment rate measures the number of employed people by the total labor force and can be calculated for cities, counties, states and countries. The average employment rate is 0.08%.

Variable cons.price.idx means consumer price index (aka CPI) which is economic indicator. It is used to measure the average changes in prices over period of time, that households pay for a basket of goods and services. Most CPI index series use 1982-1984 as the basis for comparison. [The U.S. Bureau of Labor Statistics (BLS)](https://www.bls.gov/) set the index level during this period at 100. In our dataset, the maximum CPI is approximately 94%, which means that there has been 6% decrease in the price of the market basket compared to the period of years 1982-1984.

Variable cons.conf.idx means consumer confidence index (aka CCI) which is a survey performed by monthly base. [This survey](https://tradingeconomics.com/portugal/consumer-confidence) is based on interviews with consumers about their perceptions of the current and future economic situation in the country and their tendencies to purchase. The relative value is separately computed for each question in this survey (for each question, positive responses are divided by the sum of positive and negative responses). These relative values are then compared against relative values from benchmark year 1985 ([CCI](https://en.wikipedia.org/wiki/Consumer_confidence_index)). The lowest CCI is of value -50%.


In [29]:
# Descriptive statistics for categorical objects
data.describe(include = 'object')

job  marital          education default housing   loan   contact  \
count    41188    41188              41188   41188   41188  41188     41188   
unique      12        4                  8       3       3      3         2   
top     admin.  married  university.degree      no     yes     no  cellular   
freq     10422    24928              12168   32588   21576  33950     26144   

        month day_of_week     poutcome      y  
count   41188       41188        41188  41188  
unique     10           5            3      2  
top       may         thu  nonexistent     no  
freq    13769        8623        35563  36548

From the above statistics we can see that the majority of clients work in the administrative field and have university degree. Also the most of clients are married and have borrowed money for financing their houses. The most common day of the week and month when the contact with a client was performed is Thursday and May, respectively. The majority of contacted clients have not subscribed a term deposit.

In [30]:
# Renaming columns
data.rename(columns = {'marital':'marital_status','default':'default_credit','housing':'house_loan',
                      'contact':'contact_type','duration':'contact_duration','campaign':'number_of_contacts',
                      'pdays':'days_passed','previous':'number_previous_contact','poutcome':'previous_campaign_outcome',
                      'emp.var.rate':'emp_variation_rate','cons.price.idx':'CPI','cons.conf.idx':
                      'CCI','euribor3m':'euribor_rate','nr.employed':'no_employees','y':'target'},
           inplace = True)

In [31]:
# Checking for missing values
data.isnull().sum()

age                          0
job                          0
marital_status               0
education                    0
default_credit               0
house_loan                   0
loan                         0
contact_type                 0
month                        0
day_of_week                  0
contact_duration             0
number_of_contacts           0
days_passed                  0
number_previous_contact      0
previous_campaign_outcome    0
emp_variation_rate           0
CPI                          0
CCI                          0
euribor_rate                 0
no_employees                 0
target                       0
dtype: int64

There are no missing values stored as NaN in the dataset. However, there still could be hidden missing values that we are not able to see right now. 

In [32]:
# Finding duplicate rows
duplicated_rows = data[data.duplicated()]
duplicated_rows

age          job marital_status            education default_credit  \
1266    39  blue-collar        married             basic.6y             no   
12261   36      retired        married              unknown             no   
14234   27   technician         single  professional.course             no   
16956   47   technician       divorced          high.school             no   
18465   32   technician         single  professional.course             no   
20216   55     services        married          high.school        unknown   
20534   41   technician        married  professional.course             no   
25217   39       admin.        married    university.degree             no   
28477   24     services         single          high.school             no   
32516   35       admin.        married    university.degree             no   
36951   45       admin.        married    university.degree             no   
38281   71      retired         single    university.degree             no   

      house_loan loan contact_type month day_of_week  ...  number_of_contacts  \
1266          no   no    telephone   may         thu  ...                   1   
12261         no   no    telephone   jul         thu  ...                   1   
14234         no   no     cellular   jul         mon  ...                   2   
16956        yes   no     cellular   jul         thu  ...                   3   
18465        yes   no     cellular   jul         thu  ...                   1   
20216         no   no     cellular   aug         mon  ...                   1   
20534        yes   no     cellular   aug         tue  ...                   1   
25217         no   no     cellular   nov         tue  ...                   2   
28477        yes   no     cellular   apr         tue  ...                   1   
32516        yes   no     cellular   may         fri  ...                   4   
36951         no   no     cellular   jul         thu  ...                   1   
38281         no   no    telephone   oct         tue  ...                   1   

       days_passed  number_previous_contact  previous_campaign_outcome  \
1266           999                        0                nonexistent   
12261          999                        0                nonexistent   
14234          999                        0                nonexistent   
16956          999                        0                nonexistent   
18465          999                        0                nonexistent   
20216          999                        0                nonexistent   
20534          999                        0                nonexistent   
25217          999                        0                nonexistent   
28477          999                        0                nonexistent   
32516          999                        0                nonexistent   
36951          999                        0                nonexistent   
38281          999                        0                nonexistent   

      emp_variation_rate     CPI   CCI  euribor_rate  no_employees  target  
1266                 1.1  93.994 -36.4         4.855        5191.0      no  
12261                1.4  93.918 -42.7         4.966        5228.1      no  
14234                1.4  93.918 -42.7         4.962        5228.1      no  
16956                1.4  93.918 -42.7         4.962        5228.1      no  
18465                1.4  93.918 -42.7         4.968        5228.1      no  
20216                1.4  93.444 -36.1         4.965        5228.1      no  
20534                1.4  93.444 -36.1         4.966        5228.1      no  
25217               -0.1  93.200 -42.0         4.153        5195.8      no  
28477               -1.8  93.075 -47.1         1.423        5099.1      no  
32516               -1.8  92.893 -46.2         1.313        5099.1      no  
36951               -2.9  92.469 -33.6         1.072        5076.2     yes  
38281               -3.4  92.431 -26.9         0.74

Here we can see all duplicate rows except their first occurence. These rows will be removed from the dataset.

In [33]:
# Removing duplicate rows
data.drop_duplicates(inplace = True)

In [34]:
# Checking whether the duplicate rows were correctly removed
data.duplicated().sum()

0